app/src/server/api/external/v1Api.router.ts

```
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
-              inputTokens,
-              outputTokens,
-              cost,
-            },
-          })
-          .catch((error) => captureException(error));
+        if (typedFT && typedFT.projectId !== key.projectId) {
+          throw new TRPCError({
+            message: "The model does not belong to this project",
+            code: "FORBIDDEN",
+          });
+        }
 
-        posthogServerClient?.capture({
-          distinctId: fineTune.projectId,
-          event: "fine-tune-usage",
-          properties: {
-            model: inputPayload.model,
-            inputTokens,
-            outputTokens,
-            cost,
-          },
-        });
+        try {
+          completion = await getCompletion2(typedFT, inputPayload);
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get fine-tune completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+          });
+        }
+      } else {
+        try {
+          completion = await getOpenaiCompletion(key.projectId, {
+            ...inputPayload,
+            // @ts-expect-error gotta love overloads
+            stream: input.stream,
+          });
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get OpenAI completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+            cause: error as Error,
+          });
+        }
+      }
 
-        return completion;
-      } catch (error: unknown) {
-        console.error(error);
-        throw new TRPCError({
-          message: `Failed to get completion: ${(error as Error).message}`,
-          code: "BAD_REQUEST",
+      const logRequest = ctx.headers["op-log-request"] === "true";
+      let tags: Record<string, string> = {};
+      if (ctx.headers["op-tags"]) {
+        try {
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
```

Let's actually just throw an error if you pass in an `op-tags` instance but it isn't formatted correctly. That's better than silently swallowing it and the user not knowing if they set the wrong header name, or misformatted it, or something else.

---

Also we need to use zod to validate this is actually a record of strings.